LOAD-28123 Dynamic infrastructure agent-based

guillaumebert opened the pull request:
Created from [PR#6](https://github.com/Neotys-Labs/helm-dynamic-infrastructure/pull/6) (original comments in this PR)

guillaumebert committed:
LOAD-35674 [Agent Dynamic Infra] DevOps - Helm Chart for Agent Provisioning

wlefaou committed:
LOAD-35675 - add proxy configuration to chart

Change-Id: I27f86a3bcbc6c05265aa4136e0ed09008ee2fa2b

guillaumebert committed:
Simple Review feedbacks

guillaumebert committed:
Review feedbacks - "with" and "toYaml" keywords makes it more readable

guillaumebert committed:
Review feedbacks - Updated readme (recommend not to store creds in customer values file)

guillaumebert committed:
Review feedbacks - Remove commented code

wlefaou committed:
Feedbacks - fix readme - fix custom-values comment - adapt to agent env variable changed from proxy url to host

Change-Id: I5d7ddab5ce78cef2fbe516b51e4dcf12a3970131

wlefaou committed:
Merge remote-tracking branch 'origin/featuregroup/LOAD-35674' into featuregroup/LOAD-35675

Change-Id: Id71cfca85a2eb7fc7cf8ce83c7a35b66a563d6b2

lcharlois-neotys commented on templates/agent-config.yaml:
```diff
@@ -0,0 +1,30 @@
+Prevent upgrade from old V0.x Chart to Agent-based Chart
+*/}}
+{{- if and (not $previousAgentConfigDeployment) (gt .Release.Revision 1) }}
+{{- $error := printf "ERROR Cannot upgrade.\n You are attempting to upgrade NeoLoad Helm release from chart version 0.x to 1.x.\n Please install V1.x in a new deployment.\n For detailed migration instructions, please refer to the documentation: https://github.com/Neotys-Labs/helm-dynamic-infrastructure" -}}
```
```suggestion
{{- $error := printf "ERROR Cannot upgrade.\n You are attempting to upgrade NeoLoad Helm release from chart version 0.x to 1.x.\n Please install V1.x as a new release.\n For detailed migration instructions, please refer to the documentation: https://github.com/Neotys-Labs/helm-dynamic-infrastructure" -}}
```

Better match the Helm language.

lcharlois-neotys commented on Readme.md:
```diff
@@ -14,7 +22,7 @@ In the [details chapter](#details) you can have an overview of every objects cre
 ## Installation
 
-1. Add the Neotys chart repository or update it if you already had it registered
+#### 1. Add the Neotys chart repository or update it if you already had it registered
```
```suggestion
### 1. Add the Neotys chart repository or update it if you already had it registered
```

Previous title is a 2nd level title, so this one must be a 3rd level.

lcharlois-neotys commented on Readme.md:
````diff
@@ -24,31 +32,19 @@ helm repo add neotys https://helm.prod.neotys.com/stable/
 ```
 
-2. Download and set up your **[values-custom.yaml](/values-custom.yaml)** file
+#### 2. Download and set up your **[values-custom.yaml](/values-custom.yaml)** file
````
```suggestion
### 2. Download and set up your **[values-custom.yaml](/values-custom.yaml)** file
```

Previous title is a 2nd level title, so this one must be a 3rd level.

lcharlois-neotys commented on Readme.md:
```diff
@@ -24,31 +32,19 @@ helm repo add neotys https://helm.prod.neotys.com/stable/
+Refer to the [Configuration](#Configuration) section for configuration options.
 
-4. Install with the following command
+#### 3. Install
```
```suggestion
### 3. Install
```

Previous title is a 2nd level title, so this one must be a 3rd level.

lcharlois-neotys commented on Readme.md:
```diff
@@ -59,30 +55,50 @@ $ helm uninstall my-release -n my-namespace
+| `agent.neoloadWebApiUrl`   | The URL to NeoLoad Web API (API V3, not API V4)                                                                                                                                                                                                  |                                                             |    ✅     |
+| `agent.neoloadWebApiToken` | NeoLoad Web long-term token.<br/>Recommended to set this sensitive information with commandline `--set agent.neoloadWebApiToken=TOKEN`                                                                                                           |                                                             |    ✅     |
+| `agent.name`               | Agent friendly name that will be displayed in NeoLoad Web                                                                                                                                                                                        | default name is a combination of namespace and release name |          |
+| `agent.isOpenShiftCluster` | Set to `true` if current cluster is Openshift.<br/>Set to `false` if Kubernetes                                                                                                                                                                  | `false`                                                     |          |
```
Does this really have an impact?
As I don't see any usage than declaring it and passing it to the configmap.
Is it used by the agent itself?

guillaumebert replied:
Yes, the agent itself look at all those env variables and uses it.

wlefaou committed:
Small fixes

Change-Id: I32e26b3d34ece5bb8747a688eacaed77e76bd35a

lcharlois-neotys reviewed (commented):
That looks good to me overall.
I left a few comments about layouting or wording but nothing critival.

FROD-Neotys reviewed (commented):
If I'm reading right, V1 will deploy assets from both client-based and agent-based mode.
Is it really the intention? In my mind there would be some kind of "mode" property on the chart to decide which mode to deploy.
Did you think about this?

guillaumebert commented:
> If I'm reading right, V1 will deploy assets from both client-based and agent-based mode.

No, that was not the intent. Why did you say this ?
It deploys only the assets needed for agent-based mode. And all assets are required, except the `secret.yaml` that is removed. 
We groomed and decided to increase the major version of the chart. V0 for client-based mode and V1 for agent-based mode

guillaumebert committed:
Review feedbacks wording

FROD-Neotys commented:
Unless the product choice is to deprecate client-based mode, which COULD be a rightful approach, the result of this PR is that : 

- R&D will have to maintain two distinct branches for two supported deployment mode (sure there hasn't been a lot of updates to client based mode in the recent times, but I'm talking about patterns here). It is also harder to explain to customers : go to V0 if you want client mode, V1 otherwise
- Or maybe V1 still supports client mode, but in this case I'm forced to deploy useless-to-me agent kubernetes objects, when I just want to do client mode

What do you think?

wlefaou commented on templates/agent-config.yaml:
```diff
@@ -0,0 +1,30 @@
+  KUBERNETES_NAMESPACE:  {{ .Release.Namespace | quote }}
+  RELEASE_NAME: {{ .Release.Name | quote }}
+  NLW_API_URL: {{ .Values.agent.neoloadWebApiUrl | quote }}
+  AGENT_UUID: {{ coalesce .Values.agent.uuid $previousAgentConfigDeployment.data.AGENT_UUID (randAlphaNum 32 | nospace) | quote }}
```
It does not stop at the first non null, it evaluates all arguments first, so you'll get a nil pointer at first install
Maybe could use dig like that 
```
{{ coalesce .Values.agent.uuid (dig "data" "AGENT_UUID" "" $previousAgentConfigDeployment) (randAlphaNum 32 | nospace) | quote }}
```

wlefaou replied:
by the way since we don't force uuid pattern we could add ``` | toString ``` so that 123456... will not be transformed to 1.23...e+31

guillaumebert replied:
- ok for the "dig", thanks
- I reproduced for the uuid that is a number. But this is due to you values file, you provided a number instead of a string and ``toString`` will not change it.
You must quote in ``values.yaml`` if you want a string like this `uuid: "12345678901234567890123456789012"`
instead of this `uuid: 12345678901234567890123456789012`

guillaumebert committed:
Review feedbacks - fix coalesce for the first installation

guillaumebert commented:
Discussed with @FROD-Neotys We choosed the approach of having 2 versions of this chart to maintain on 2 different branches instead of having 1 chart that can deploy one or another depending on a property.
Reasons:
- customers will mostly deploy agent-based providers as the recommended (which is the latest version)
- codebase is cleaner and don't need any {{if agent-mode / else ...}}
- it force migration and consequently removing unnecessary objects (`secrets.yaml`) that decrease security only required by client-based mode and that decrease security

guillaumebert commented on templates/agent-config.yaml:
```diff
@@ -0,0 +1,30 @@
+Prevent upgrade from old V0.x Chart to Agent-based Chart
+*/}}
+{{- if and (not $previousAgentConfigDeployment) (gt .Release.Revision 1) }}
+{{- $error := printf "ERROR Cannot upgrade.\n You are attempting to upgrade NeoLoad Helm release from chart version 0.x to 1.x.\n Please install V1.x as a new release.\n For detailed migration instructions, please refer to the documentation: https://github.com/Neotys-Labs/helm-dynamic-infrastructure" -}}
```
The link goes to the public github repo readme. This is already a stable link for me

sylbn commented on values.yaml:
```diff
@@ -13,4 +13,30 @@ registryKey:
+    drop:
+      - ALL
+  allowPrivilegeEscalation: false
+  runAsNonRoot: true
```
I think this will not work on openshift

guillaumebert replied:
Cannot set explicitly ``runAsUser`` because on Openshift clusters, the allowed users are a range and the value `1000` will prevent the pod to start

guillaumebert commented on templates/agent.yaml:
```diff
@@ -0,0 +1,46 @@
+        {{- end }}
+    spec:
+      serviceAccountName: {{ .Release.Name }}-sa
+      securityContext:
+        {{- toYaml .Values.podSecurityContext | nindent 8 }}
```
I define podSecurityContext in values.yaml

guillaumebert committed:
AI Review feedbacks

wlefaou committed:
Merge remote-tracking branch 'origin/featuregroup/LOAD-35674' into featuregroup/LOAD-35675

Change-Id: I0881dd9f15721c2eb206d93d57b85865cb0f1ed4

guillaumebert committed:
Change Chart version to 1.0.0-alpha.1

wlefaou committed:
Merge pull request #7 from Neotys-Labs/featuregroup/LOAD-35675

LOAD-35675 add agent proxy configuration

guillaumebert committed:
Demo feedbacks

- Set strategy to Recreate (and put readme note to prevent having multiple replicas)
- add documentation for `image.agent` (to use custom docker image)
- add documentation of labels and annotations
- split annotations between pods and deployment

guillaumebert committed:
fix

guillaumebert committed:
Merge pull request #9 from Neotys-Labs/featuregroup/demo-feedbacks

Demo feedbacks

guillaumebert committed:
Remove alpha version

guillaumebert committed:
Allow setting a custom pull secret to pull agent image (#11)

wlefaou committed:
Fix NPE when configuring a proxy without credentials

Change-Id: If4c1e725169891399940eebf1bfa215cc9d670eb

wlefaou committed:
Fix NPE when configuring a proxy without credentials

Fix NPE when configuring a proxy without credentials

wlefaou committed:
Add possibility to pass environment variables and mount volumes - add documentation on how to handle self-signed proxy certificates (#13)
